src/dex/maverick-v2/maverick-v2.ts

```diff
@@ -0,0 +1,392 @@
+  Logger,
+  NumberAsString,
+  DexExchangeParam,
+} from '../../types';
+import { SwapSide, Network, NULL_ADDRESS } from '../../constants';
+import * as CALLDATA_GAS_COST from '../../calldata-gas-cost';
+import { getBigIntPow, getDexKeysWithNetwork, isTruthy } from '../../utils';
+import { Context, IDex } from '../../dex/idex';
+import { IDexHelper } from '../../dex-helper/idex-helper';
+import { MaverickV2Data, PoolAPIResponse } from './types';
+import { SimpleExchange } from '../simple-exchange';
+import {
+  MaverickV2Config,
+  MAV_V2_BASE_GAS_COST,
+  MAV_V2_TICK_GAS_COST,
+  MAVERICK_API_URL,
+} from './config';
+import { MaverickV2EventPool } from './maverick-v2-pool';
+import { SUBGRAPH_TIMEOUT } from '../../constants';
+import { Interface } from '@ethersproject/abi';
+import MaverickV2PoolABI from '../../abi/maverick-v2/MaverickV2Pool.json';
+import ERC20ABI from '../../abi/erc20.json';
+import { extractReturnAmountPosition } from '../../executor/utils';
+const EFFICIENCY_FACTOR = 3;
+
+export class MaverickV2 extends SimpleExchange implements IDex<MaverickV2Data> {
+  pools: { [key: string]: MaverickV2EventPool } = {};
+  readonly hasConstantPriceLargeAmounts = false;
+  readonly needWrapNative = true;
+
+  readonly isFeeOnTransferSupported = false;
+
+  public static dexKeysWithNetwork: { key: string; networks: Network[] }[] =
+    getDexKeysWithNetwork(MaverickV2Config);
+
+  logger: Logger;
+
+  public static erc20Interface = new Interface(ERC20ABI);
+
+  constructor(
+    readonly network: Network,
+    readonly dexKey: string,
+    readonly dexHelper: IDexHelper,
+    protected adapters = {},
+    protected config = MaverickV2Config[dexKey][network],
+    protected maverickV2Iface = new Interface(MaverickV2PoolABI),
+  ) {
+    super(dexHelper, dexKey);
+    this.logger = dexHelper.getLogger(dexKey);
+  }
+
+  async initializePricing(blockNumber: number) {
+    await this.setupEventPools(blockNumber);
+  }
+
+  async setupEventPools(blockNumber: number) {
+    const pools = (await this._queryPoolsAPI(SUBGRAPH_TIMEOUT))?.pools;
+
+    await Promise.all(
+      pools?.map(async (pool: any) => {
+        const eventPool = new MaverickV2EventPool(
+          this.dexKey,
+          this.network,
+          this.dexHelper,
+          this.logger,
+          {
+            address: pool.tokenA.address,
+            symbol: pool.tokenA.symbol,
+            decimals: pool.tokenA.decimals,
+          },
+          {
+            address: pool.tokenB.address,
+            symbol: pool.tokenB.symbol,
+            decimals: pool.tokenB.decimals,
+          },
+          BigInt(pool.fee * 1e6) * BigInt(1e12),
+          BigInt(pool.feeB * 1e6) * BigInt(1e12),
+          BigInt(pool.tickSpacing),
+          BigInt(0),
+          BigInt(pool.lookback),
+          BigInt(pool.lowerTick),
+          pool.id,
+          this.config.poolLensAddress,
+        );
+        const onChainState = await eventPool.generateState(blockNumber);
```

Comment:
I believe its a good place to use multicall here to fetch states for all pools together